File: backend/src/haan/utils/cost_tracker.py
"""
Token usage and cost tracking for the Haan.ai backend.

Tracks per-task and daily token usage and estimated costs across all
LLM providers. Supports budget limits that raise BudgetExceededError
when exceeded, allowing the pipeline to stop gracefully.

Usage:
    from haan.utils.cost_tracker import cost_tracker

    cost_tracker.set_budget(per_task=5.0, daily=50.0)
    cost_tracker.add_usage("claude-sonnet-4-20250514", input_tokens=1000, output_tokens=500)
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any


class BudgetExceededError(Exception):
    """Raised when a budget limit (per-task or daily) is exceeded."""

    pass


# Approximate cost per 1M tokens (USD) for common models
# These are rough estimates used for budget tracking, not billing.
MODEL_COSTS: dict[str, dict[str, float]] = {
    # Anthropic
    "claude-sonnet-4-20250514": {"input": 3.0, "output": 15.0},
    "claude-opus-4-20250514": {"input": 15.0, "output": 75.0},
    "claude-haiku-3-20250306": {"input": 0.25, "output": 1.25},
    # OpenAI
    "gpt-4o": {"input": 2.5, "output": 10.0},
    "gpt-4o-mini": {"input": 0.15, "output": 0.6},
    "o3-mini": {"input": 1.1, "output": 4.4},
    # Google
    "gemini-2.5-pro": {"input": 1.25, "output": 10.0},
    "gemini-2.5-flash": {"input": 0.15, "output": 0.6},
}

# Fallback cost for unknown models
DEFAULT_COST = {"input": 3.0, "output": 15.0}


@dataclass
class UsageRecord:
    """A single usage record for a model invocation."""

    model: str
    input_tokens: int
    output_tokens: int
    estimated_cost: float
    timestamp: float = field(default_factory=time.time)


@dataclass
class CostSummary:
    """Summary of token usage and estimated costs."""

    total_input_tokens: int = 0
    total_output_tokens: int = 0
    total_cost: float = 0.0
    records: list[UsageRecord] = field(default_factory=list)


class CostTracker:
    """
    Tracks token usage and estimated costs across all LLM calls.

    Maintains both per-task and daily running totals, and enforces
    configurable budget limits.
    """

    def __init__(self) -> None:
        self._task_summary = CostSummary()
        self._daily_summary = CostSummary()
        self._daily_reset_time: float = self._get_day_start()
        self._budget_per_task: float = 0.0  # 0 = unlimited
        self._budget_daily: float = 0.0  # 0 = unlimited

    def set_budget(
        self,
        per_task: float = 0.0,
        daily: float = 0.0,
    ) -> None:
        """
        Set budget limits in USD.

        Args:
            per_task: Maximum cost per pipeline task (0 = unlimited).
            daily: Maximum daily cost across all tasks (0 = unlimited).
        """
        self._budget_per_task = per_task
        self._budget_daily = daily

    def add_usage(
        self,
        model: str,
        input_tokens: int,
        output_tokens: int,
    ) -> float:
        """
        Record token usage for a model call.

        Args:
            model: Model identifier (e.g., "claude-sonnet-4-20250514").
            input_tokens: Number of input tokens consumed.
            output_tokens: Number of output tokens generated.

        Returns:
            Estimated cost in USD for this call.

        Raises:
            BudgetExceededError: If adding this usage exceeds a budget limit.
        """
        # Check if we need to reset daily totals
        self._maybe_reset_daily()

        # Calculate cost
        costs = MODEL_COSTS.get(model, DEFAULT_COST)
        estimated_cost = (
            (input_tokens / 1_000_000) * costs["input"]
            + (output_tokens / 1_000_000) * costs["output"]
        )

        record = UsageRecord(
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            estimated_cost=estimated_cost,
        )

        # Update task totals
        self._task_summary.total_input_tokens += input_tokens
        self._task_summary.total_output_tokens += output_tokens
        self._task_summary.total_cost += estimated_cost
        self._task_summary.records.append(record)

        # Update daily totals
        self._daily_summary.total_input_tokens += input_tokens
        self._daily_summary.total_output_tokens += output_tokens
        self._daily_summary.total_cost += estimated_cost
        self._daily_summary.records.append(record)

        # Check budget limits
        if self._budget_per_task > 0 and self._task_summary.total_cost > self._budget_per_task:
            raise BudgetExceededError(
                f"Per-task budget exceeded: ${self._task_summary.total_cost:.4f} "
                f"> ${self._budget_per_task:.2f}"
            )
        if self._budget_daily > 0 and self._daily_summary.total_cost > self._budget_daily:
            raise BudgetExceededError(
                f"Daily budget exceeded: ${self._daily_summary.total_cost:.4f} "
                f"> ${self._budget_daily:.2f}"
            )

        return estimated_cost

    def reset_task_cost(self) -> None:
        """Reset per-task cost tracking (called at start of each pipeline run)."""
        self._task_summary = CostSummary()

    def get_task_summary(self) -> dict[str, Any]:
        """Return a summary dict of the current task's costs."""
        return {
            "input_tokens": self._task_summary.total_input_tokens,
            "output_tokens": self._task_summary.total_output_tokens,
            "estimated_cost": round(self._task_summary.total_cost, 6),
            "num_calls": len(self._task_summary.records),
        }

    def get_daily_summary(self) -> dict[str, Any]:
        """Return a summary dict of today's cumulative costs."""
        self._maybe_reset_daily()
        return {
            "input_tokens": self._daily_summary.total_input_tokens,
            "output_tokens": self._daily_summary.total_output_tokens,
            "estimated_cost": round(self._daily_summary.total_cost, 6),
            "num_calls": len(self._daily_summary.records),
        }

    def _maybe_reset_daily(self) -> None:
        """Reset daily totals if we've crossed into a new day."""
        current_day_start = self._get_day_start()
        if current_day_start > self._daily_reset_time:
            self._daily_summary = CostSummary()
            self._daily_reset_time = current_day_start

    @staticmethod
    def _get_day_start() -> float:
        """Return the Unix timestamp for the start of the current UTC day."""
        now = time.time()
        return now - (now % 86400)


# Module-level singleton
cost_tracker = CostTracker()

File: src/utils/cost-tracker.ts
import { eventBus } from './event-bus.js';
import { logger } from './logger.js';

interface ModelPricing {
  inputPer1M: number;
  outputPer1M: number;
}

export class BudgetExceededError extends Error {
  constructor(
    public readonly limit: string,
    public readonly current: number,
    public readonly max: number,
  ) {
    super(`Budget exceeded (${limit}): $${current.toFixed(4)} / $${max.toFixed(4)}`);
    this.name = 'BudgetExceededError';
  }
}

interface BudgetConfig {
  dailyLimit?: number;
  perTaskLimit?: number;
}

const PRICING: Record<string, ModelPricing> = {
  // Anthropic
  'claude-opus-4-5-20251101': { inputPer1M: 5.0, outputPer1M: 25.0 },
  'claude-opus-4-5': { inputPer1M: 5.0, outputPer1M: 25.0 },
  'claude-sonnet-4-5-20250929': { inputPer1M: 3.0, outputPer1M: 15.0 },
  'claude-sonnet-4-5': { inputPer1M: 3.0, outputPer1M: 15.0 },
  // OpenAI
  'gpt-4.1': { inputPer1M: 2.0, outputPer1M: 8.0 },
  'gpt-4o': { inputPer1M: 2.5, outputPer1M: 10.0 },
  'gpt-4.1-mini': { inputPer1M: 0.4, outputPer1M: 1.6 },
  'gpt-5': { inputPer1M: 5.0, outputPer1M: 20.0 },
  'gpt-5-mini': { inputPer1M: 1.0, outputPer1M: 4.0 },
  'gpt-5.2-codex': { inputPer1M: 5.0, outputPer1M: 20.0 },
  'gpt-5.1-codex': { inputPer1M: 3.0, outputPer1M: 12.0 },
  'gpt-5.1-codex-max': { inputPer1M: 5.0, outputPer1M: 20.0 },
  'o4-mini': { inputPer1M: 1.1, outputPer1M: 4.4 },
  'o3-mini': { inputPer1M: 1.1, outputPer1M: 4.4 },
  // Google
  'gemini-2.5-pro': { inputPer1M: 1.25, outputPer1M: 10.0 },
  'gemini-3-pro-preview': { inputPer1M: 1.25, outputPer1M: 10.0 },
  'gemini-2.5-flash': { inputPer1M: 0.15, outputPer1M: 0.6 },
};

interface UsageEntry {
  model: string;
  inputTokens: number;
  outputTokens: number;
  cost: number;
}

class CostTracker {
  private entries: UsageEntry[] = [];
  private totalCost = 0;
  private totalInputTokens = 0;
  private totalOutputTokens = 0;
  private taskCost = 0;
  private dailyCost = 0;
  private dailyResetDate: string = new Date().toISOString().slice(0, 10);
  private budget: BudgetConfig = {};

  constructor() {
    eventBus.on('cost:update', (data) => this.track(data));
  }

  /**
   * Set budget limits. Pass undefined values to disable a limit.
   */
  setBudget(config: BudgetConfig): void {
    this.budget = config;
    logger.info('cost-tracker', `Budget set: daily=$${config.dailyLimit ?? 'unlimited'}, perTask=$${config.perTaskLimit ?? 'unlimited'}`);
  }

  /**
   * Reset per-task cost tracking (call at the start of each pipeline).
   */
  resetTaskCost(): void {
    this.taskCost = 0;
  }

  /**
   * Check if a new LLM call with estimated cost would exceed the budget.
   * Throws BudgetExceededError if it would.
   */
  checkBudget(estimatedCost: number = 0): void {
    // Rotate daily cost if date changed
    const today = new Date().toISOString().slice(0, 10);
    if (today !== this.dailyResetDate) {
      this.dailyCost = 0;
      this.dailyResetDate = today;
    }

    if (this.budget.perTaskLimit !== undefined) {
      if (this.taskCost + estimatedCost > this.budget.perTaskLimit) {
        throw new BudgetExceededError('per-task', this.taskCost, this.budget.perTaskLimit);
      }
    }

    if (this.budget.dailyLimit !== undefined) {
      if (this.dailyCost + estimatedCost > this.budget.dailyLimit) {
        throw new BudgetExceededError('daily', this.dailyCost, this.budget.dailyLimit);
      }
    }
  }

  track(entry: { model: string; inputTokens: number; outputTokens: number; cost?: number }) {
    const pricing = PRICING[entry.model] || { inputPer1M: 5.0, outputPer1M: 15.0 };
    const calculated = (
      (entry.inputTokens / 1_000_000) * pricing.inputPer1M +
      (entry.outputTokens / 1_000_000) * pricing.outputPer1M
    );
    const baseCost = (entry.cost !== undefined && entry.cost > 0) ? entry.cost : calculated;
    const cost = baseCost * 1.5; // 1.5x markup for profit margin
    const record: UsageEntry = { ...entry, cost };
    this.entries.push(record);
    this.totalCost += cost;
    this.taskCost += cost;
    this.dailyCost += cost;
    this.totalInputTokens += entry.inputTokens;
    this.totalOutputTokens += entry.outputTokens;
  }

  getSummary() {
    return {
      totalCost: this.totalCost,
      totalInputTokens: this.totalInputTokens,
      totalOutputTokens: this.totalOutputTokens,
      totalTokens: this.totalInputTokens + this.totalOutputTokens,
      entries: this.entries,
      byModel: this.entries.reduce<Record<string, { input: number; output: number; cost: number }>>((acc, e) => {
        if (!acc[e.model]) acc[e.model] = { input: 0, output: 0, cost: 0 };
        acc[e.model].input += e.inputTokens;
        acc[e.model].output += e.outputTokens;
        acc[e.model].cost += e.cost;
        return acc;
      }, {}),
    };
  }

  formatCost(): string {
    return `$${this.totalCost.toFixed(4)}`;
  }

  formatTokens(): string {
    const total = this.totalInputTokens + this.totalOutputTokens;
    if (total > 1_000_000) return `${(total / 1_000_000).toFixed(1)}M`;
    if (total > 1_000) return `${(total / 1_000).toFixed(1)}K`;
    return `${total}`;
  }

  reset() {
    this.entries = [];
    this.totalCost = 0;
    this.totalInputTokens = 0;
    this.totalOutputTokens = 0;
  }
}

export const costTracker = new CostTracker();
